File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n=s.size();
        int l=0,r=0;
        int maxlen=0; 
        unordered_map<char,int>mp;
        while(r<n){
            mp[s[r]]++;
            while(mp[s[r]]>1){
                mp[s[l]]--;
                l++;
            }
            maxlen=max(maxlen,r-l+1);
            r++;
        }
        return maxlen;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {

//         int[] freq = new int[256];

//         int left=0;
//         int max=0;
        
//         for(int right=0;right<s.length();right++){
//             char c=s.charAt(right);
//             freq[c]++;
                
//     while (freq[c] > 1) {
//     freq[s.charAt(left)]--;
//     left++;

//         }
//         max = Math.max(max, right - left + 1);
     
// }
// return max;

 HashSet<Character> set = new HashSet<>();

    int max=0;
    int left=0;

    for(int right=0;right<s.length();right++){
        char c=s.charAt(right);

        while(set.contains(c)){
            set.remove(s.charAt(left));
            left++;
        }
        set.add(c);
        max = Math.max(max, right - left + 1);

    }
    return max;


        
    }
}